ctru-rs/src/services/reference.rs

```diff
@@ -0,0 +1,44 @@
+use crate::Error;
+use std::sync::Mutex;
+pub(crate) struct ServiceReference {
+    counter: &'static Mutex<usize>,
+    close: Box<dyn Fn()>,
+}
+
+impl ServiceReference {
+    pub fn new<S, E>(
+        counter: &'static Mutex<usize>,
+        allow_multiple: bool,
+        start: S,
+        close: E,
+    ) -> crate::Result<Self>
+    where
+        S: FnOnce() -> crate::Result<()>,
+        E: Fn() + 'static,
+    {
+        let mut value = counter.lock().unwrap(); // todo: handle poisoning
```

Comment:
Do you want to handle the poisoning TODOs in this PR?

Reply:
I’d like to. What can we do about it, though?

Reply:
We can either ignore it, unwrap/panic, or handle it. In our case, what it means to be poisoned is that the lock may have been held during a panic, which dropped the lock. So, maybe the start function panics, for example. Or, start doesn't panic, but somehow there's a panic between start and the counter increment.

If we ignore it, that is probably fine, but we just need to make sure we can't panic between the start task succeeding and the counter getting incremented. This is probably fine. Unwrapping/panicking is also an option, and would be the safest, plus it's unlikely that the lock gets poisoned in the first place, so this is fine to do. Handling the poison error seems hard because that implies fixing the broken state. But it's hard to know what's broken. So not much to do here I think.

I may be missing a few things since this is all off the top of my head.

Reply:
I guess I'll change from an `unwrap` to an expect with a bit of info, but other than that it's hard. Still, panicking the thread while holding the lock is pretty much impossible, so I don't think there are issues.